pkg/services/ngalert/state/manager.go

```diff
@@ -314,41 +312,79 @@ func (st *Manager) flushState(ctx context.Context) {
 	st.cache.mtxStates.Lock()
 	defer st.cache.mtxStates.Unlock()
 	totalStates, errorsCnt := 0, 0
+	var stateSlice []*State
 	for _, orgStates := range st.cache.states {
 		for _, ruleStates := range orgStates {
+			stateSlice = stateSlice[:0]
 			for _, state := range ruleStates {
-				err := st.saveState(ctx, state)
-				totalStates++
-				if err != nil {
-					st.log.Error("failed to save alert state", append(state.GetRuleKey().LogContext(), "labels", state.Labels.String(), "state", state.State.String(), "err", err.Error()))
-					errorsCnt++
-				}
+				stateSlice = append(stateSlice, state)
+			}
+			savedCount, failedCount := st.saveAlertStates(ctx, stateSlice...)
+			if failedCount != 0 {
+				st.log.Error("failed to save alert states.", "count", failedCount)
 			}
+			errorsCnt += failedCount
+			totalStates += savedCount
 		}
 	}
+
 	st.log.Info("the state has been flushed", "total_instances", totalStates, "errors", errorsCnt, "took", st.clock.Since(t))
 }
 
-func (st *Manager) saveState(ctx context.Context, s *State) error {
-	cmd := ngModels.SaveAlertInstanceCommand{
-		RuleOrgID:         s.OrgID,
-		RuleUID:           s.AlertRuleUID,
-		Labels:            ngModels.InstanceLabels(s.Labels),
-		State:             ngModels.InstanceStateType(s.State.String()),
-		StateReason:       s.StateReason,
-		LastEvalTime:      s.LastEvaluationTime,
-		CurrentStateSince: s.StartsAt,
-		CurrentStateEnd:   s.EndsAt,
+// TODO: Is the `State` type necessary? Should it embed the instance?
```

Comment:
State is a kinda domain model decoupled from the DTO model... I would like us to have that separation for other models too. It will help adopt different storages (like S3). The conversion should be on storage level, though. Not in this PR.